homeassistant/auth/providers/command_line.py

```diff
@@ -0,0 +1,163 @@
+}, extra=vol.PREVENT_EXTRA)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class InvalidAuthError(HomeAssistantError):
+    """Raised when authentication with given credentials fails."""
+
+
+@AUTH_PROVIDERS.register("command_line")
+class CommandLineAuthProvider(AuthProvider):
+    """Auth provider validating credentials by calling a command."""
+
+    DEFAULT_TITLE = "Command Line Authentication"
+
+    # which keys to accept from a program's stdout
+    ALLOWED_META_KEYS = ("name",)
+
+    def __init__(self, *args: Any, **kwargs: Any) -> None:
+        """Extend parent's __init__.
+
+        Adds self._user_meta dictionary to hold the user-specific
+        attributes provided by external programs.
+        """
+        super().__init__(*args, **kwargs)
+        self._user_meta = {}  # type: Dict[str, Dict[str, Any]]
+
+    async def async_login_flow(self, context: Optional[dict]) -> LoginFlow:
+        """Return a flow to login."""
+        return CommandLineLoginFlow(self)
+
+    async def async_validate_login(self, username: str, password: str) -> None:
+        """Validate a username and password."""
+        env = {
+            "username": username,
+            "password": password,
+        }
+        try:
+            # pylint: disable=no-member
+            process = await asyncio.subprocess.create_subprocess_exec(
+                self.config[CONF_COMMAND], *self.config[CONF_ARGS],
+                env=env,
+                stdout=asyncio.subprocess.PIPE
+                if self.config[CONF_META] else None,
+            )
+            stdout, _ = (await process.communicate())
+        except OSError as err:
+            # happens when command doesn't exist or permission is denied
+            _LOGGER.error("Error while authenticating %r: %s",
+                          username, err)
+            raise InvalidAuthError
+
+        if process.returncode != 0:
+            _LOGGER.error("User %r failed to authenticate, command exited "
+                          "with code %d.",
+                          username, process.returncode)
+            raise InvalidAuthError
+
+        if self.config[CONF_META]:
+            meta = {}  # type: Dict[str, str]
+            for _line in stdout.splitlines():
+                try:
+                    line = _line.decode().lstrip()
+                    if line.startswith("#"):
+                        continue
+                    key, value = line.split("=", 1)
+                except ValueError:
+                    # malformed line
+                    continue
+                key = key.strip()
+                value = value.strip()
+                if key in self.ALLOWED_META_KEYS:
+                    meta[key] = value
+            self._user_meta[username] = meta
+
+    async def async_get_or_create_credentials(
+            self, flow_result: Dict[str, str]
+    ) -> Credentials:
+        """Get credentials based on the flow result."""
+        username = flow_result["username"]
```

Comment:
Should we strip the username?

(usernames are not stripped before being passed on because we didn't do it initially, and so doing so now will be a breaking change. We have put a guard in place to ensure it won't happen and we will start stripping usernames in 2020)